src/AppInstallerCommonCore/Public/winget/Pin.h

```diff
@@ -6,63 +6,116 @@
-        // The package is blocked from 'upgrade --all' and 'upgrade <package>'.
-        // User has to unblock to allow update.
-        Blocking,
         // The package is excluded from 'upgrade --all', unless '--include-pinned' is added.
         // 'upgrade <package>' is not blocked.
         Pinning,
         // The package is pinned to a specific version range.
         Gating,
+        // The package is blocked from 'upgrade --all' and 'upgrade <package>'.
+        // User has to unblock to allow update.
+        Blocking,
     };
 
     std::string_view ToString(PinType type);
     PinType ConvertToPinTypeEnum(std::string_view in);
 
-    // The set of values needed to uniquely identify a Pin
+    // Determines which of two pin types is more strict.
+    PinType StrictestPinType(PinType first, PinType second);
+
+    // The type of an extra string used for identifying an installed app for pinning
+    enum class ExtraIdStringType
+    {
+        None,
+        ProductCode,
+        PackageFamilyName,
+    };
+
+    std::string_view ToString(ExtraIdStringType type);
+
+    // The set of values needed to uniquely identify a Pin.
+    // We use both a PackageId and a ProductCode or PackageFamilyName
+    // to be able to deal with packages that match with multiple installed apps.
+    // This helps us deal better with apps that we don't match properly elsewhere.
     struct PinKey
     {
         PinKey() {}
-        PinKey(const Manifest::Manifest::string_t& packageId, std::string_view sourceId)
-            : PackageId(packageId), SourceId(sourceId) {}
+        PinKey(const Manifest::Manifest::string_t& packageId, std::string_view sourceId, ExtraIdStringType extraIdType, std::string_view extraId = {})
+            : PackageId(packageId), SourceId(sourceId), ExtraIdType(extraIdType), ExtraId(extraId) {}
 
         bool operator==(const PinKey& other) const
         {
-            return PackageId == other.PackageId && SourceId == other.SourceId;
+            return PackageId == other.PackageId
+                && SourceId == other.SourceId
+                && ExtraIdType == other.ExtraIdType
+                && ExtraId == other.ExtraId;
         }
+
         bool operator!=(const PinKey& other) const
         {
             return !(*this == other);
         }
+
         bool operator<(const PinKey& other) const
         {
-            return PackageId < other.PackageId || (PackageId == other.PackageId && SourceId < other.SourceId);
+            if (PackageId != other.PackageId)
+            {
+                return PackageId < other.PackageId;
+            }
+            else if (SourceId != other.SourceId)
+            {
+                return SourceId < other.SourceId;
+            }
+            else if (ExtraIdType != other.ExtraIdType)
+            {
+                return ExtraIdType < other.ExtraIdType;
+            }
+            else
+            {
+                return ExtraId < other.ExtraId;
+            }

```

Comment:
```suggestion
            // std::tie implements tuple comparison, wherein it checks the first item in the tuple,
            // iff the first elements are equal, then the second element is used for comparison, and so on
            return std::tie(PackageId, SourceId, ExtraIdType, ExtraId) < 
            	std::tie(other.PackageId, other.SourceId, other.ExtraIdType, other.ExtraId);
```
https://bajamircea.github.io/coding/cpp/2017/03/10/std-tie.html
https://en.cppreference.com/w/cpp/utility/tuple/operator_cmp